libpod/in_memory_state.go

```diff
@@ -0,0 +1,279 @@
+package libpod
+
+import (
+	"github.com/docker/docker/pkg/truncindex"
+	"github.com/kubernetes-incubator/cri-o/pkg/registrar"
+	"github.com/pkg/errors"
+)
+
+// An InMemoryState is a purely in-memory state store
+type InMemoryState struct {
+	pods         map[string]*Pod
+	containers   map[string]*Container
+	podNameIndex *registrar.Registrar
+	podIDIndex   *truncindex.TruncIndex
+	ctrNameIndex *registrar.Registrar
+	ctrIDIndex   *truncindex.TruncIndex
+}
+
+// NewInMemoryState initializes a new, empty in-memory state
+func NewInMemoryState() (State, error) {
+	state := new(InMemoryState)
+
+	state.pods = make(map[string]*Pod)
+	state.containers = make(map[string]*Container)
+
+	state.podNameIndex = registrar.NewRegistrar()
+	state.ctrNameIndex = registrar.NewRegistrar()
+
+	state.podIDIndex = truncindex.NewTruncIndex([]string{})
+	state.ctrIDIndex = truncindex.NewTruncIndex([]string{})
+
+	return state, nil
+}
+
+// GetContainer retrieves a container from its full ID
+func (s *InMemoryState) GetContainer(id string) (*Container, error) {
+	if id == "" {
+		return nil, ErrEmptyID
+	}
+
+	ctr, ok := s.containers[id]
+	if !ok {
+		return nil, errors.Wrapf(ErrNoSuchCtr, "no container with ID %s found", id)
+	}
+
+	return ctr, nil
+}
+
+// LookupContainer retrieves a container by full ID, unique partial ID, or name
+func (s *InMemoryState) LookupContainer(idOrName string) (*Container, error) {
+	if idOrName == "" {
+		return nil, ErrEmptyID
+	}
+
+	fullID, err := s.ctrNameIndex.Get(idOrName)
+	if err != nil {
+		if err == registrar.ErrNameNotReserved {
+			// What was passed is not a name, assume it's an ID
+			fullID, err = s.ctrIDIndex.Get(idOrName)
+			if err != nil {
+				if err == truncindex.ErrNotExist {
+					return nil, errors.Wrapf(ErrNoSuchCtr, "no container found with name or ID %s", idOrName)
+				} else {
+					return nil, errors.Wrapf(err, "error performing truncindex lookup for ID %s", idOrName)
+				}
+			}
+		} else {
+			return nil, errors.Wrapf(err, "error performing registry lookup for ID %s", idOrName)
+		}
+	}
+
+	ctr, ok := s.containers[fullID]
+	if !ok {
+		// This should never happen
+		return nil, errors.Wrapf(ErrInternal, "mismatch in container ID registry and containers map for ID %s", fullID)
+	}
+
+	return ctr, nil
+}
+
+// HasContainer checks if a container with the given ID is present in the state
+func (s *InMemoryState) HasContainer(id string) (bool, error) {
```

Comment:
I think this should just be boolean.  No errors.  Although not sure why you would call this versus just calling GetContainer?

Reply:
If we ever move to database-backed or file-backed storage, all of these could result in errors (to access the database, misformatted data, etc). So I'd like to keep the error in there, even if we don't use it